src/backend/distributed/planner/extended_op_node_utils.c

```diff
@@ -42,15 +46,25 @@ static bool ShouldPullDistinctColumn(bool repartitionSubquery,
  * value should be used in a read-only manner.
  */
 ExtendedOpNodeProperties
-BuildExtendedOpNodeProperties(MultiExtendedOp *extendedOpNode, bool
-							  pullUpIntermediateRows)
+BuildExtendedOpNodeProperties(MultiExtendedOp *extendedOpNode,
+							  bool pullUpIntermediateRows)
 {
 	ExtendedOpNodeProperties extendedOpNodeProperties;
 
 	List *tableNodeList = FindNodesOfType((MultiNode *) extendedOpNode, T_MultiTable);
 	bool groupedByDisjointPartitionColumn =
+		!extendedOpNode->hasNonPushableWindowFunction &&
 		GroupedByPartitionColumn((MultiNode *) extendedOpNode, extendedOpNode);
 
+	if (groupedByDisjointPartitionColumn)
+	{
+		/*
+		 * pullUpIntermediate rows tells us aggregates can't be split across
+		 * worker & master. Prefer to push down to worker if possible.
+		 */
+		pullUpIntermediateRows = false;
```

Comment:
I feel like the naming of pullUpIntermediateRows could be improved, especially because it is not the same thing as `extendedOpNodeProperties.pullUpIntermediateRows`. It sort of means `maybePullUpIntermediateRowsUnlessGroupingByPartitionColumn` right now, but something like `hasNonDistributableAggregates` would clarify things.